### PR TITLE
Draft: [onert] Support RoPE operation

### PR DESCRIPTION
**Pull request by ys44kim**
This commit supports RoPE Operation for onert

ONE-DCO-1.0-Signed-off-by: youngsik kim <ys44.kim@samsung.com>

### CODE REVIEW

File: compute/cker/include/cker/Types.h [hunk 1]
```diff
@@ -63,6 +63,12 @@ enum class ComparisonOpType
   LessEqual
 };
 
+enum class RoPEMode
+{
+  kGptNeox = 0,
+  kGptJ = 1,
+};
+
```
Comment (ys44kim) [inline compute/cker/include/cker/Types.h:71]: It was added to support two modes (GPT-NEOX, GPT-J) of RoPE.
Currently, only GPT-NEOX is supported.

File: compute/cker/include/cker/operation/RoPE.h [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+inline void RoPE(const RoPEMode mode, const Shape &input_shape, const T *input_data,
+                 const Shape &sin_table_shape, const T *sin_table_data,
+                 const Shape &cos_table_shape, const T *cos_table_data, const Shape &output_shape,
+                 T *output_data)
+{
+  if (input_shape.Dims(3) != sin_table_shape.Dims(3))
+    throw std::runtime_error("the dimension(3) of input and sin_table do not match");
+
+  if (input_shape.Dims(3) != cos_table_shape.Dims(3))
+    throw std::runtime_error("the dimension(3) of input and cos_table do not match");
```
Comment (ys44kim) [inline compute/cker/include/cker/operation/RoPE.h:39]: if the sin/cos table and input dim(3) are different, an error occurs in the following operation 
and it is processed as throw.

File: runtime/onert/backend/cpu/KernelGenerator.cc [hunk 3]
```diff
@@ -1543,6 +1544,26 @@ void KernelGenerator::visit(const ir::operation::LSTM &node)
...
+  const auto sin_table{node.getInputs().at(ir::operation::RoPE::Input::SIN_TABLE)};
+  const auto cos_table{node.getInputs().at(ir::operation::RoPE::Input::COS_TABLE)};
+  const auto output_index{node.getOutputs().at(ir::operation::RoPE::Output::OUTPUT)};
+
+  auto mode = ops::getRoPEMode(node.param().mode);
+
+  auto input_tensor = _tensor_reg->getPortableTensor(input_index);
+  auto sin_tensor = _tensor_reg->getPortableTensor(sin_table);
+  auto cos_tensor = _tensor_reg->getPortableTensor(cos_table);
+  auto output_tensor = _tensor_reg->getPortableTensor(output_index);
```
Comment (ys44kim) [inline runtime/onert/backend/cpu/KernelGenerator.cc:1559]: The current implementation has three inputs(input, sin_table, and cos_table)
so, input the sin/cos table at the upper layer (ex.app).

File: compute/cker/include/cker/operation/RoPE.h [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+            output_data[offset] = x0 * cos_table_data[i3] - x1 * sin_table_data[i3];
+            output_data[offset + i3_n / 2] =
+              x0 * sin_table_data[i3 + i3_n / 2] + x1 * cos_table_data[i3 + i3_n / 2];
+          }
+        }
+      }
+    }
+  }
+  else
```
Comment (ys44kim) [inline compute/cker/include/cker/operation/RoPE.h:68]: ### BUILD_TYPE = Release 

#### before fuse
$ ./Product/x86_64-linux.release/out/bin/onert_run ~/model/rope/Net_RoPE_000.circle   
```
===================================
MODEL_LOAD   takes 0.251 ms
PREPARE      takes 3.557 ms
EXECUTE      takes 0.036 ms
- MEAN     :  0.036 ms
- MAX      :  0.036 ms
- MIN      :  0.036 ms
- GEOMEAN  :  0.036 ms
===================================
```
#### after fuse
$ ./Product/x86_64-linux.release/out/bin/onert_run ~/model/rope/RoPE_000.circle   
```
===================================
MODEL_LOAD   takes 0.169 ms
PREPARE      takes 2.384 ms
EXECUTE      takes 0.020 ms
- MEAN     :  0.020 ms
- MAX      :  0.020 ms
- MIN      :  0.020 ms
- GEOMEAN  :  0.020 ms
===================================
```

**Comment by ys44kim** [comment]
@glistening  @hseok-oh @ragmani @zetwhite
this pr is draft of rope in onert
please review it.

**Comment by ys44kim** [comment]
@nnfw-bot test-all

**Comment by hseok-oh** [comment]
I have fundamental question about RoPE operation spec. On your implementation, RoPE gets sin & cos table as input. But IMHO, that looks very specific workaround to support RoPE operation with less performance loss on DSP/NPU which is not having sin/cos calculation unit. 
So I don't think we need to require sin/cos table as input, and it should be handled on compiler (or total toolchain) level.

@glistening Could you give opinion about this?

**Comment by jinevening** [comment]
> that looks very specific workaround to support RoPE operation with less performance loss 

`LlamaAttention` module in `transformers` package is transitioning to get sin/cos table (`position_embeddings`) as inputs. So, IMHO the implementation is not that backend-specific.
(https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L346-L352)

> So I don't think we need to require sin/cos table as input, and it should be handled on compiler (or total toolchain) level.

I agree that sin/cos table should be generated within a toolchain. I guess that something placed higher than the current onert can handle that.

**Comment by hseok-oh** [comment]
>`LlamaAttention` module in `transformers` package is transitioning to get sin/cos table (`position_embeddings`) as inputs. So, IMHO the implementation is not that backend-specific.

I didn't know that. Thank you explanation.

**Comment by ys44kim** [comment]
> So I don't think we need to require sin/cos table as input, and it should be handled on compiler (or total toolchain) level.

> I agree that sin/cos table should be generated within a toolchain. I guess that something placed higher than the current onert can handle that.

@hseok-oh @jinevening 
if sin/cos table is generated within a toolchain, how to usually implement it?
is the shape of the rope op changed? or is it changed a implementation?
Actually i understood it as input because I thought backend does not make sin / cos table.

**Comment by jinevening** [comment]
> if sin/cos table is generated within a toolchain, how to usually implement it?

@ys44kim  You can find how sin/cos table is created in the below source code.
- https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L146-L164

It may be possible to (1) implement the above code with circle operators, or (2) just let runtime handle those stuffs. I imagined that we're going to take the second approach (please correct me if I'm wrong @glistening @hseok-oh ). So, the current implementation of rope don't have to be changed IIUC.

**Comment by ys44kim** [comment]
this PR is merged with other PRs.

**Pull request closed without merging**